pkg/dataobj/internal/dataset/value_encoding_plain.go

```diff
@@ -240,26 +241,56 @@ func (dec *plainBytesDecoder) Decode(s []Value) (int, error) {
 		} else if err != nil {
 			return i, err
 		}
-		s[i] = v
 	}
 	return len(s), nil
 }
 
 // decode decodes a string.
-func (dec *plainBytesDecoder) decode() (Value, error) {
+func (dec *plainBytesDecoder) decode(v *Value) error {
 	sz, err := binary.ReadUvarint(dec.r)
 	if err != nil {
-		return ByteArrayValue([]byte{}), err
+		return err
 	}
 
-	dst := make([]byte, int(sz))
+	dst := getDestinationBuffer(v, int(sz))
 	if _, err := io.ReadFull(dec.r, dst); err != nil {
-		return ByteArrayValue([]byte{}), err
+		return err
 	}
-	return ByteArrayValue(dst), nil
+
+	v.num = sz
+	v.any = (bytearray)(unsafe.SliceData(dst))
+	return nil
 }
 
 // Reset implements [valueDecoder]. It resets the decoder to read from r.
 func (dec *plainBytesDecoder) Reset(r streamio.Reader) {
 	dec.r = r
 }
+
+func getDestinationBuffer(v *Value, sz int) []byte {
+	if v.cap == 0 {
+		dst := make([]byte, sz)
+		v.any = (bytearray)(unsafe.SliceData(dst))
+		v.cap = uint64(cap(dst))
+		return dst
+	}
+
+	var dst []byte
+	// Depending on which type this value was previously used for dictates how we reference the memory.
+	switch v.any.(type) {
+	case stringptr:
+		dst = unsafe.Slice(v.any.(stringptr), int(v.cap))
```

Comment:
I'm concerned with us doing this for `stringptr`:

If a caller passes in a string Value created via StringValue, its memory would get overwritten here and violate Go's memory model, which can cause bugs that are very difficult to track down, if the program doesn't segfault outright. 

IMO, if it's a stringptr, we should consider eating the cost and allocating a new slice, just to save ourselves from dealing with obscure memory corruption bugs. (And then we should also avoid using string values wherever possible)

Reply:
I agree. I've added a commit to remove string types. It's a big change so this PR will be XXL, but it should be self-contained.
Without changing the dataobj.Record to stop using labels.Labels, it doesn't have a big impact on the capacities that need to be maintained. logs.Decode is simpler, though, because the internal logs.Record type can be changed to use `[]byte` instead of `string` for values.
Let me know what you think. I'm happy to revert this commit and tackle it in a future PR if that would be clearer!